src/adaptation/precond.jl

```diff
@@ -111,72 +117,185 @@ function get_cov(wc::WelfordCov{T})::Matrix{T} where {T<:AbstractFloat}
+mutable struct DiagPreconditioner{T<:Real,AT<:AbstractVector{T}} <: AbstractPreconditioner
     n_min   :: Int
     ve  :: VarEstimator{T}
     var :: AT
 end
 
-function DiagPreConditioner(d::Int, n_min::Int=10)
+function DiagPreconditioner(d::Int, n_min::Int=10)
     ve = WelfordVar(0, zeros(d), zeros(d))
-    return DiagPreConditioner(n_min, ve, Vector(ones(d)))
+    return DiagPreconditioner(n_min, ve, Vector(ones(d)))
 end
 
-function string(dpc::DiagPreConditioner)
+function string(dpc::DiagPreconditioner)
     return string(dpc.var)
 end
 
-function adapt!(dpc::DiagPreConditioner, θ::AbstractVector{<:Real}, α::AbstractFloat, is_update::Bool=true)
+function adapt!(dpc::DiagPreconditioner, θ::AbstractVector{<:Real}, α::AbstractFloat, is_update::Bool=true)
+    if length(θ) > length(dpc.var)
+        @assert dpc.ve.n == 0 "Cannot resize a var estimator when it contains samples."
+        dpc.ve = WelfordVar(0, zeros(length(θ)), zeros(length(θ)))
+        dpc.var = zeros(length(θ))
+    end
     add_sample!(dpc.ve, θ)
     if dpc.ve.n >= dpc.n_min && is_update
         dpc.var .= get_var(dpc.ve)
     end
 end
 
-reset!(dpc::DiagPreConditioner) = reset!(dpc.ve)
+reset!(dpc::DiagPreconditioner) = reset!(dpc.ve)
 
-function getM⁻¹(dpc::DiagPreConditioner)
+function getM⁻¹(dpc::DiagPreconditioner)
     return dpc.var
 end
 
-mutable struct DensePreConditioner{T<:AbstractFloat} <: AbstractPreConditioner
+mutable struct DensePreconditioner{T<:AbstractFloat} <: AbstractPreconditioner
     n_min :: Int
     ce    :: CovEstimator{T}
     covar :: Matrix{T}
 end
 
-function DensePreConditioner(d::Integer, n_min::Int=10)
+function DensePreconditioner(d::Integer, n_min::Int=10)
     ce = WelfordCov(d)
     # TODO: take use of the line below when we have an interface to set which pre-conditioner to use
     # ce = NaiveCov()
-    return DensePreConditioner(n_min, ce, LinearAlgebra.diagm(0 => ones(d)))
+    return DensePreconditioner(n_min, ce, LinearAlgebra.diagm(0 => ones(d)))
 end
 
-function string(dpc::DensePreConditioner)
+function string(dpc::DensePreconditioner)
     return string(LinearAlgebra.diag(dpc.covar))
 end
 
-function adapt!(dpc::DensePreConditioner, θ::AbstractVector{<:AbstractFloat}, α::AbstractFloat, is_update::Bool=true)
+function adapt!(dpc::DensePreconditioner, θ::AbstractVector{<:AbstractFloat}, α::AbstractFloat, is_update::Bool=true)
+    if length(θ) > size(dpc.covar,1)
+        @assert dpc.ce.n == 0 "Cannot resize a var estimator when it contains samples."
+        dpc.ce = WelfordCov(length(θ))
+        dpc.covar = zeros(length(θ),length(θ))
+    end
     add_sample!(dpc.ce, θ)
     if dpc.ce.n >= dpc.n_min && is_update
         dpc.covar .= get_cov(dpc.ce)
     end
 end
 
-reset!(dpc::DensePreConditioner) = reset!(dpc.ce)
+reset!(dpc::DensePreconditioner) = reset!(dpc.ce)
 
-function getM⁻¹(dpc::DensePreConditioner)
+function getM⁻¹(dpc::DensePreconditioner)
     return dpc.covar
 end
+
+####
```

Comment:
Why do we move the code here? Does it make AHMC dependent on Adaptation then?

Reply:
I think ideally we would like AHMC without the Adaptation model will work as well.

Reply:
I think we can replace `Metric` types with `PDMat` (see https://github.com/TuringLang/AdvancedHMC.jl/issues/34), so this doesn't matter in the future.

Reply:
Also, AHMC depends on Adaption anyway: some core features like step-size adaption is impossible without Adaption.

Reply:
But shouldn't a call of `sample()` without adaptation is expected to work?

Reply:
I see your point - but I don't think this matters a lot.  Besides, so this dependency will go away if we fix #34, right?

Reply:
Yes that's right! I'm fine with the current code then.